w-near/src/lib.rs

```diff
@@ -0,0 +1,1351 @@
+    /// The number of allowances
+    pub num_allowances: u32,
+}
+
+impl Account {
+    /// Initializes a new Account with 0 balance and no allowances for a given `account_hash`.
+    pub fn new(account_hash: Vec<u8>) -> Self {
+        Self { balance: 0, allowances: LookupMap::new(account_hash), num_allowances: 0 }
+    }
+
+    /// Sets allowance for account `escrow_account_id` to `allowance`.
+    pub fn set_allowance(&mut self, escrow_account_id: &AccountId, allowance: Balance) {
+        let escrow_hash = env::sha256(escrow_account_id.as_bytes());
+        if allowance > 0 {
+            if self.allowances.insert(&escrow_hash, &allowance).is_none() {
+                self.num_allowances += 1;
+            }
+        } else {
+            if self.allowances.remove(&escrow_hash).is_some() {
+                self.num_allowances -= 1;
+            }
+        }
+    }
+
+    /// Returns the allowance of account `escrow_account_id`.
+    pub fn get_allowance(&self, escrow_account_id: &AccountId) -> Balance {
+        let escrow_hash = env::sha256(escrow_account_id.as_bytes());
+        self.allowances.get(&escrow_hash).unwrap_or(0)
+    }
+}
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FungibleToken {
+    /// sha256(AccountID) -> Account details.
+    pub accounts: LookupMap<Vec<u8>, Account>,
+
+    /// Total supply of the all token.
+    pub total_supply: Balance,
+}
+
+impl Default for FungibleToken {
+    fn default() -> Self {
+        env::panic(b"Contract should be initialized before usage.")
+    }
+}
+
+#[near_bindgen]
+impl FungibleToken {
+    #[init]
+    pub fn new() -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            accounts: LookupMap::new(b"a".to_vec()),
+            total_supply: 0u128
+        }
+    }
+
+    /// Deposit NEAR and send wNear tokens to the predecessor account
+    /// Requirements:
+    /// * `amount` must be a positive integer
+    /// * Caller of the method has to attach deposit enough to cover:
+    ///   * The `amount` of wNear tokens being minted, and
+    ///   * The storage difference at the fixed storage price defined in the contract.
+    #[payable]
+    pub fn deposit(&mut self, amount: U128) {
+        // Proxy through to deposit_to() making the receiver_id the predecessor
+        self.deposit_to(env::predecessor_account_id(), amount);
+    }
+
+    /// Deposit NEAR from the predecessor account and send wNear to a specific receiver_id
+    /// Requirements:
+    /// * `receiver_id` cannot be this contract
+    /// * `receiver_id` must be a valid account Id
+    /// * `amount` must be a positive integer
+    /// * Caller of the method has to attach deposit enough to cover:
+    ///   * The `amount` of wNear tokens being minted, and
+    ///   * The storage difference at the fixed storage price defined in the contract.
+    #[payable]
+    pub fn deposit_to(&mut self, receiver_id: AccountId, amount: U128) {
```

Comment:
With `near-sdk = 2.0.0` crate, you can use `ValidAccountId` argument type from `near_sdk::json_types` mod.
It's a wrapped `AccountId` that verifies account ID validity at deserialization time, so you don't need to check `env:: is_valid_account_id` later. 
Then you can cast it to `AccountId` using `into`. 

```suggestion
    pub fn deposit_to(&mut self, receiver_id: ValidAccountId, amount: U128) {
        let receiver_id: AccountId = receiver_id.into();
```

Reply:
@evgenykuzyakov I have replaced `AccountId` where I could to `ValidAccountId`. However, `deposit_to()` is called by `deposit` and the first argument passed is `env::predecessor_account_id()` which returns a string. As far as I can see, there is no way to cast from `AccountId` to `ValidAccountId`. Therefore, for `deposit_to()` and `withdraw_to()`, I have left the param type as `AccountId` for `receiver_id`

Reply:
@evgenykuzyakov In the end I had to revert all the param changes I did back to all being of type `AccountId`. I couldn't find a way of creating a `ValidAccountId` from a string which meant that all my tests were failing. I am happy to use `ValidAccountId` if I can do the conversion from string. Let me know what you think